packages/dxf-deserializer/translate.js

```diff
@@ -0,0 +1,880 @@
+        }
+        break
+      case 'mesh':
+// FIXME complete translation
+        console.log('##### mesh')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        //objects.push(instantiateMesh(obj, layers, options))
+        translateMesh(obj, layers, options)
+        break
+
+      // 2D or 3D entities
+      case 'arc':
+        console.log('##### arc')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translateArc(obj, layers, options)
+        break
+      case 'circle':
+        console.log('##### circle')
+        current = translateCurrent(current, layers, parts, options)
+        translateCircle(obj, layers, options)
+        parts = []
+        break
+      case 'ellipse':
+        console.log('##### ellipse')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translateEllipse(obj, layers, options)
+        break
+      case 'line':
+        console.log('##### line')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translateLine(obj, layers, options)
+        break
+      case 'polyline':
+        if (current === null) {
+          console.log('##### start of polyline')
+          current = getPolyType(obj)
+          current['name'] = 'jscad' + numobjs
+          numobjs = numobjs + 1
+        }
+        break
+      case 'vertex':
+        //console.log('##### vertex')
+        p = instantiateVector(obj)
+        break
+      case 'seqend':
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        break
+
+      // 2D entities
+      case 'lwpolyline':
+        console.log('##### lwpolyline')
+        current = translateCurrent(current, layers, parts, options)
+        parts = []
+        translatePath2D(obj, layers, options)
+        break
+
+      default:
+        console.log('##### ERROR')
+        console.log(obj.type)
+        break
+    }
+    // accumlate polygons if necessary
+    if (p instanceof CSG.Polygon) {
+      // console.log('##### push Polygon')
+      parts.push(p)
+    }
+    // accumlate vectors if necessary
+    if (p instanceof CSG.Vector3D) {
+      // console.log('##### push Vector3D')
+      parts.push(p)
+    }
+    if (p instanceof CSG.Vector2D) {
+      // console.log('##### push Vector2D')
+      parts.push(p)
+    }
+  }
+  // translate the last object if necessary
+  current = translateCurrent(current, layers, parts, options)
+
+  // debug output
+  // console.log('**************************************************')
+  let script = 'function main() {\n  let layers = []\n  return layers.concat('
+  layers.forEach(
+    function (layer) {
+      let name = layer['name'] || 'Unknown'
+      script += 'layer' + name + '(),'
+    }
+  )
+  script += '[])\n}\n'
+
+  // add helper functions for polygons and lines
+  script += 'function createVertex(point) {\n'
```

Comment:
it would be better to use es6 string templates (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) instead of string concatenation whenever possible

Reply:
Seems that MS Explorer doesn’t implement templates. Maybe it’s still too early to switch. Thoughts?

Reply:
we have loooong since given up support for IE(failing webgl support in there in the first place), Edge supports them : remember : the past is the past :)

Reply:
OK! We e given ya all there is captain! Warp 10!

Reply:
I’ll make the changes tomorrow. :)